content/blog/2025-12-16-ChocoPy2Bril.md

```diff
@@ -0,0 +1,97 @@
+
+For the implementation, we directly called the "yangdanny97/chocopy-python-compiler" code to generate the AST. Luckily, the AST that was generated was basically in JSON format. As a result, we just had to hard code the translation between the ChocoPy AST nodes and the Bril instructions. Most notably, we created a `CodeState` class to keep track of temporary variables and variable types. This helps generate unique temporary variables names during the translation. The `translate_prog` function serves to handle function definitions, the ChocoPy main function body, as well as any global variable declarations. `translate_func_def` is responsible for mapping arguments and local variable definitions, translating statements, and adding a default return value if none is proveded. `translate_var_def` translates variable definitions, and `translate_expr` handles the translation of instructions, including literals, identifiers, binary operations, unary operations, and function calls. Finally, `translate_stmt` translates statements such as assignment, if statements, while loops, return statements, and expression statements including print. The main function orchestrates the overall AST translation, including calling "yangdanny97/chocopy-python-compiler" to create the ChocoPy AST, as well as calling the appropriate functions from above to generate the Bril code.
+
+Finally, testing also played an important role in our implementation. We curated a test suite of 20 unit tests that covered the various features we implemented in our ChocoPy2Bril translator. Each test case consisted of a ChocoPy program, and we wrote a script that verified that the result of running the generated Bril code matched the ChocoPy program's output. This helped us ensure the correctness of our implementation and catch any bugs or edge cases. Our test suites were partially adapted/inspired from the test suites provided in the "yangdanny97/chocopy-python-compiler" repository. Initially, we wanted to directly use tests from "yangdanny97/chocopy-python-compiler". However, we soon realized that most of those tests were fairly complex and covered a lot of ChocoPy features that we were not planning on implementing. As a result, we had to edit them to fit our implemented features, as well as manually write some of our tests from scratch. Additionally, on the GitHub issue for this project, we were intially planning on using Turnt for testing. However, we soon realized that since the way that we were checking correctness/calling functions differed across test cases, it was simply easier to write our own testing script that handled each test case individually.
```

Comment:
> However, we soon realized that since the way that we were checking correctness/calling functions differed across test cases

I don't quite understand this: isn't the correctness check always "the compiled Bril program prints the same thing as the original ChocoPy program"? In that case, what are the relevant differences? Could you maybe say a little bit more about the nuances of setting these things up to test them?